tools/evaluate_queries/ranked.cpp

```diff
@@ -0,0 +1,81 @@
+#include "def.hpp"
+#include "query/algorithm/ranked_and_query.hpp"
+#include "query/algorithm/ranked_or_query.hpp"
+
+namespace pisa {
+
+#define PISA_RANKED_OR_QUERY_LOOP(SCORER, INDEX, WAND)                                 \
+    template <>                                                                        \
+    auto query_loop<BOOST_PP_CAT(INDEX, _index),                                       \
+                    pisa::ranked_or_query,                                             \
+                    wand_data<WAND>,                                                   \
+                    SCORER<wand_data<WAND>>>(BOOST_PP_CAT(INDEX, _index) const &index, \
+                                             wand_data<WAND> const &,                  \
+                                             SCORER<wand_data<WAND>> scorer,           \
+                                             std::vector<Query> const &queries,        \
+                                             int k)                                    \
+        ->std::vector<ResultVector>                                                    \
+    {                                                                                  \
+        std::vector<ResultVector> results(queries.size());                             \
+        auto run = ranked_or_query(k);                                                 \
+        for (std::size_t qidx = 0; qidx < queries.size(); ++qidx) {                    \
+            auto query = queries[qidx];                                                \
+            auto cursors = make_scored_cursors(index, scorer, query);                  \
+            run(gsl::make_span(cursors), index.num_docs());                            \
+            results[qidx] = run.topk();                                                \
+        }                                                                              \
+        return results;                                                                \
+    }
+
+#define LOOP_BODY(R, DATA, T)                                \
+    PISA_RANKED_OR_QUERY_LOOP(bm25, T, wand_data_raw)        \
+    PISA_RANKED_OR_QUERY_LOOP(dph, T, wand_data_raw)         \
+    PISA_RANKED_OR_QUERY_LOOP(pl2, T, wand_data_raw)         \
+    PISA_RANKED_OR_QUERY_LOOP(qld, T, wand_data_raw)         \
+    PISA_RANKED_OR_QUERY_LOOP(bm25, T, wand_data_compressed) \
+    PISA_RANKED_OR_QUERY_LOOP(dph, T, wand_data_compressed)  \
+    PISA_RANKED_OR_QUERY_LOOP(pl2, T, wand_data_compressed)  \
+    PISA_RANKED_OR_QUERY_LOOP(qld, T, wand_data_compressed)  \
+/**/
+BOOST_PP_SEQ_FOR_EACH(LOOP_BODY, _, PISA_INDEX_TYPES);
+#undef LOOP_BODY
+#undef PISA_RANKED_OR_QUERY_LOOP
+
+#define PISA_RANKED_AND_QUERY_LOOP(SCORER, INDEX, WAND)                                \
```

Comment:
Why is this a macro? Can't be a function?

Reply:
It could but, imo, it ends up being messier because you still have to define explicit specialization to compile in the same unit. And so it looks something like this:

1. Introduce a function template for AND:

```cpp
template <typename Index, typename Scorer>
auto query_loop_and(Index const &index, Scorer scorer, std::vector<Query> const &queries, int k)
    -> std::vector<ResultVector>
{
    std::vector<ResultVector> results(queries.size());
    auto run = ranked_and_query(k);
    for (std::size_t qidx = 0; qidx < queries.size(); ++qidx) {
        auto query = queries[qidx];
        auto cursors = make_scored_cursors(index, scorer, query);
        run(gsl::make_span(cursors), index.num_docs());
        results[qidx] = run.topk();
    }
    return results;
}
```
2. Explicit instantiation.
```cpp
#define LOOP_BODY(R, DATA, T)                                                     \
    template <>                                                                   \
    auto query_loop_and<BOOST_PP_CAT(T, _index), bm25<wand_data<wand_data_raw>>>( \
        BOOST_PP_CAT(T, _index) const &index,                                     \
        bm25<wand_data<wand_data_raw>> scorer,                                    \
        std::vector<Query> const &queries,                                        \
        int k)                                                                    \
        ->std::vector<ResultVector>;                                              \
    /* REPEATED FOR ALL (scorer, wand) */
/**/
BOOST_PP_SEQ_FOR_EACH(LOOP_BODY, _, PISA_INDEX_TYPES);
#undef LOOP_BODY
```

Note that the last part you need to repeat 8 times, and each time the scorer and wand data are in two different places far from each other; it's much more difficult to maintain.

Unless you have yet another idea how to do this...